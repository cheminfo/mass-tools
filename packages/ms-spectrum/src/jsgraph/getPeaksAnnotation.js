'use strict';

/**
 *
 * @param {array} bestPeaks
 * @param {object} [options={}]
 * @param {array} [options.mfColors={}]
 * @param {number} [options.numberDigits=5]
 * @param {number} [options.numberMFs=0]
 * @param {boolean} [options.showMF=false]
 * @param {array} [options.mfColors={}]
 * @param {number} [options.charge=1]
 * @param {number} [options.shift=0]
 * @param {object} [options.mfPrefs]
 * @param {number} [options.displayCharge=true]
 * @param {number} [options.displayProperties=[]] Array of properties name to display
 * @returns {Promise}
 */

async function getPeaksAnnotation(bestPeaks, options = {}) {
  const emdb = new (require('emdb'))();

  options = Object.assign({ limit: 5, precision: 100 }, options);

  let {
    numberDigits = 5,
    shift = 0,
    showMF = false,
    numberMFs = 0,
    charge = 1,
    mfPrefs = {},
    displayCharge = true,
    displayProperties = [],
    mfColors = [
      { limit: 3, color: 'green' },
      { limit: 20, color: 'lightgreen' },
      { limit: 50, color: 'lightorange' },
    ],
  } = options;
  if (showMF && !numberMFs) numberMFs = 1;
  let annotations = [];
  bestPeaks.sort((a, b) => (a.close ? -1 : b.close ? 1 : 0));

  for (let peak of bestPeaks) {
    let textLine = 0;
    let annotation;
    if (peak.close) {
      annotation = {
        type: 'line',
        _highlight: peak._highlight,
        info: peak,
        position: [
          {
            y: peak.y,
            dy: '-5px',
            x: peak.x,
          },
          {
            y: peak.y,
            dy: '-15px',
            x: peak.x,
          },
        ],
      };
      annotations.push(annotation);
      annotation = {
        type: 'ellipse',
        _highlight: peak._highlight,
        info: peak,
        position: [
          {
            y: peak.y,
            dy: '-15px',
            x: peak.x,
          },
        ],
        props: {
          rx: '3px',
          ry: '3px',
          fillOpacity: 0.3,
        },
      };
    } else {
      annotation = {
        type: 'line',
        _highlight: peak._highlight,
        position: [
          {
            y: peak.y,
            dy: '-5px',
            x: peak.x,
          },
          {
            y: peak.y,
            dy: '-25px',
            x: peak.x,
          },
        ],
        labels: [
          {
            text: (peak.x + shift).toFixed(numberDigits),
            color: 'red',
            position: {
              x: peak.x,
              y: peak.y,
              dy: `${textLine++ * -13 - 17}px`,
              dx: '2px',
            },
          },
        ],
      };

      if (displayCharge) {
        annotation.labels.push({
          text: `Z:${peak.charge}`,
          color: 'grey',
          position: {
            x: peak.x,
            y: peak.y,
            dy: '-4px',
            dx: '2px',
          },
        });
      }

      if (numberMFs) {
        // we have 2 cases. Either there is a shift and we deal with differences
        // otherwise it is absolute
        // if there is a shift we consider only a neutral loss and the parameter charge is important
        if (shift) {
          // neutral loss
          let currentMfPrefs = Object.assign({}, mfPrefs, {
            allowNeutral: true,
            ionizations: '',
          });
          // we need to deal with the precision and increase it
          currentMfPrefs.precision =
            (currentMfPrefs.precision / Math.max(Math.abs(peak.x + shift), 1)) *
            peak.x;
          await emdb.fromMonoisotopicMass(
            Math.abs((peak.x + shift) * charge),
            currentMfPrefs,
          );
        } else {
          await emdb.fromMonoisotopicMass(Math.abs(peak.x * charge), mfPrefs);
        }

        let mfs = emdb.get('monoisotopic');
        let numberOfMFS = Math.min(mfs.length, numberMFs);

        for (let i = 0; i < numberOfMFS; i++) {
          let mf = mfs[i];

          let ppm = shift ? (mf.ms.ppm / shift) * mfs[0].ms.em : mf.ms.ppm;
          annotation.labels.push({
            text: mf.mf,
            color: getColor(mfColors, Math.abs(ppm)),
            position: {
              x: peak.x,
              y: peak.y,
              dy: `${textLine++ * -13 - 17}px`,
              dx: '2px',
            },
          });
        }
      }

      if (displayProperties.length > 0) {
        for (let property of displayProperties) {
          annotation.labels.push({
            text: peak[property],
            color: 'red',
            position: {
              x: peak.x,
              y: peak.y,
              dy: `${textLine++ * -13 - 17}px`,
              dx: '2px',
            },
          });
        }
      }
    }
    annotations.push(annotation);
  }
  return annotations;
}

function getColor(colors, value) {
  for (let color of colors) {
    if (value < color.limit) return color.color;
  }
  return 'lightgrey';
}

module.exports = getPeaksAnnotation;
