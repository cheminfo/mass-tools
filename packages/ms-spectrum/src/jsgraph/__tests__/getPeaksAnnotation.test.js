import { describe, expect, it } from 'vitest';

import { getPeaksAnnotation } from '../getPeaksAnnotation';

describe('test getPeaksAnnotation', () => {
  const peaks = [
    { x: 12, y: 1 },
    { x: 24, y: 2 },
    { x: 36, y: 3 },
    { x: 42, y: 4, charge: 1 },
  ];

  it('default options', async () => {
    let result = await getPeaksAnnotation(peaks);

    expect(result).toMatchSnapshot();
  });

  it('shift -12', async () => {
    let result = await getPeaksAnnotation(peaks, {
      shift: -11.99999,
      showMF: true,
      charge: 2,
    });

    expect(result).toMatchSnapshot();
  });

  it('cap numberMFs to 2', async () => {
    let result = await getPeaksAnnotation(peaks, {
      numberMFs: 2,
    });

    expect(result).toMatchSnapshot();
  });
});
