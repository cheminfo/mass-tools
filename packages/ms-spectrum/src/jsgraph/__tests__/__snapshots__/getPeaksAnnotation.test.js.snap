// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test getPeaksAnnotation default options 1`] = `
Array [
  Object {
    "_highlight": undefined,
    "labels": Array [
      Object {
        "color": "red",
        "position": Object {
          "dx": "2px",
          "dy": "-17px",
          "x": 12,
          "y": 1,
        },
        "text": "12.00000",
      },
      Object {
        "color": "grey",
        "position": Object {
          "dx": "2px",
          "dy": "-4px",
          "x": 12,
          "y": 1,
        },
        "text": "Z:undefined",
      },
    ],
    "position": Array [
      Object {
        "dy": "-5px",
        "x": 12,
        "y": 1,
      },
      Object {
        "dy": "-25px",
        "x": 12,
        "y": 1,
      },
    ],
    "type": "line",
  },
  Object {
    "_highlight": undefined,
    "labels": Array [
      Object {
        "color": "red",
        "position": Object {
          "dx": "2px",
          "dy": "-17px",
          "x": 24,
          "y": 2,
        },
        "text": "24.00000",
      },
      Object {
        "color": "grey",
        "position": Object {
          "dx": "2px",
          "dy": "-4px",
          "x": 24,
          "y": 2,
        },
        "text": "Z:undefined",
      },
    ],
    "position": Array [
      Object {
        "dy": "-5px",
        "x": 24,
        "y": 2,
      },
      Object {
        "dy": "-25px",
        "x": 24,
        "y": 2,
      },
    ],
    "type": "line",
  },
  Object {
    "_highlight": undefined,
    "labels": Array [
      Object {
        "color": "red",
        "position": Object {
          "dx": "2px",
          "dy": "-17px",
          "x": 36,
          "y": 3,
        },
        "text": "36.00000",
      },
      Object {
        "color": "grey",
        "position": Object {
          "dx": "2px",
          "dy": "-4px",
          "x": 36,
          "y": 3,
        },
        "text": "Z:undefined",
      },
    ],
    "position": Array [
      Object {
        "dy": "-5px",
        "x": 36,
        "y": 3,
      },
      Object {
        "dy": "-25px",
        "x": 36,
        "y": 3,
      },
    ],
    "type": "line",
  },
  Object {
    "_highlight": undefined,
    "labels": Array [
      Object {
        "color": "red",
        "position": Object {
          "dx": "2px",
          "dy": "-17px",
          "x": 42,
          "y": 4,
        },
        "text": "42.00000",
      },
      Object {
        "color": "grey",
        "position": Object {
          "dx": "2px",
          "dy": "-4px",
          "x": 42,
          "y": 4,
        },
        "text": "Z:undefined",
      },
    ],
    "position": Array [
      Object {
        "dy": "-5px",
        "x": 42,
        "y": 4,
      },
      Object {
        "dy": "-25px",
        "x": 42,
        "y": 4,
      },
    ],
    "type": "line",
  },
]
`;

exports[`test getPeaksAnnotation shift -12 1`] = `
Array [
  Object {
    "_highlight": undefined,
    "labels": Array [
      Object {
        "color": "red",
        "position": Object {
          "dx": "2px",
          "dy": "-17px",
          "x": 12,
          "y": 1,
        },
        "text": "0.00001",
      },
      Object {
        "color": "grey",
        "position": Object {
          "dx": "2px",
          "dy": "-4px",
          "x": 12,
          "y": 1,
        },
        "text": "Z:undefined",
      },
    ],
    "position": Array [
      Object {
        "dy": "-5px",
        "x": 12,
        "y": 1,
      },
      Object {
        "dy": "-25px",
        "x": 12,
        "y": 1,
      },
    ],
    "type": "line",
  },
  Object {
    "_highlight": undefined,
    "labels": Array [
      Object {
        "color": "red",
        "position": Object {
          "dx": "2px",
          "dy": "-17px",
          "x": 24,
          "y": 2,
        },
        "text": "12.00001",
      },
      Object {
        "color": "grey",
        "position": Object {
          "dx": "2px",
          "dy": "-4px",
          "x": 24,
          "y": 2,
        },
        "text": "Z:undefined",
      },
    ],
    "position": Array [
      Object {
        "dy": "-5px",
        "x": 24,
        "y": 2,
      },
      Object {
        "dy": "-25px",
        "x": 24,
        "y": 2,
      },
    ],
    "type": "line",
  },
  Object {
    "_highlight": undefined,
    "labels": Array [
      Object {
        "color": "red",
        "position": Object {
          "dx": "2px",
          "dy": "-17px",
          "x": 36,
          "y": 3,
        },
        "text": "24.00001",
      },
      Object {
        "color": "grey",
        "position": Object {
          "dx": "2px",
          "dy": "-4px",
          "x": 36,
          "y": 3,
        },
        "text": "Z:undefined",
      },
    ],
    "position": Array [
      Object {
        "dy": "-5px",
        "x": 36,
        "y": 3,
      },
      Object {
        "dy": "-25px",
        "x": 36,
        "y": 3,
      },
    ],
    "type": "line",
  },
  Object {
    "_highlight": undefined,
    "labels": Array [
      Object {
        "color": "red",
        "position": Object {
          "dx": "2px",
          "dy": "-17px",
          "x": 42,
          "y": 4,
        },
        "text": "30.00001",
      },
      Object {
        "color": "grey",
        "position": Object {
          "dx": "2px",
          "dy": "-4px",
          "x": 42,
          "y": 4,
        },
        "text": "Z:undefined",
      },
    ],
    "position": Array [
      Object {
        "dy": "-5px",
        "x": 42,
        "y": 4,
      },
      Object {
        "dy": "-25px",
        "x": 42,
        "y": 4,
      },
    ],
    "type": "line",
  },
]
`;
